src/Compilers/CSharp/Portable/Binder/Lifetime.cs

```diff
@@ -0,0 +1,129 @@
+        /// Since sibling scopes do not intersect and a value cannot escape from one to another without
+        /// escaping to a wider scope, we can use simple depth numbering without ambiguity.
+        /// </summary>
+        private readonly uint _value;
+        private Lifetime(uint value) => _value = value;
+
+        /// <summary>
+        /// The "calling method" scope that is outside of the containing method/lambda.
+        /// If something can escape to this scope, it can escape to any scope in a given method through a ref parameter or return.
+        /// </summary>
+        public static Lifetime CallingMethod => new Lifetime(CallingMethodRaw);
+
+        /// <summary>
+        /// The "return-only" scope that is outside of the containing method/lambda.
+        /// If something can escape to this scope, it can escape to any scope in a given method or can be returned, but it can't escape through a ref parameter.
+        /// </summary>
+        public static Lifetime ReturnOnly => new Lifetime(ReturnOnlyRaw);
+
+        /// <summary>
+        /// The "current method" scope that is just inside the containing method/lambda.
+        /// If something can escape to this scope, it can escape to any scope in a given method, but cannot be returned.
+        /// </summary>
+        public static Lifetime CurrentMethod => new Lifetime(CurrentMethodRaw);
+
+        /// <summary>
+        /// Gets a lifetime which is "empty". i.e. which refers to a variable whose storage is never allocated.
+        /// </summary>
+        public static Lifetime Empty => new Lifetime(uint.MaxValue);
+
+        /// <summary>
+        /// Gets a lifetime which is narrower than the given lifetime.
+        /// Used to "enter" a nested local scope.
+        /// </summary>
+        public Lifetime Narrower()
+        {
+            var result = new Lifetime(this._value + 1);
+            // Narrower() operator should always result in a local lifetime
+            Debug.Assert(!result.IsReturnable);
+            return result;
+        }
+
+        /// <summary>
+        /// Gets a lifetime which is wider than the given lifetime.
+        /// Used to "exit" a nested local scope.
+        /// </summary>
+        public Lifetime Wider()
+        {
+            // Wider() operator should always start from a local lifetime
+            Debug.Assert(!this.IsReturnable);
+            return new Lifetime(this._value - 1);
+        }
+
+        public bool IsCallingMethod => _value == CallingMethodRaw;
+        public bool IsReturnOnly => _value == ReturnOnlyRaw;
+        public bool IsReturnable => _value is CallingMethodRaw or ReturnOnlyRaw;
+
+        /// <summary>Returns true if a 'ref' with this lifetime can be converted to the 'other' lifetime. Otherwise, returns false.</summary>
+        /// <remarks>Generally, a wider lifetime is convertible to a narrower lifetime.</remarks>
+        public bool IsConvertibleTo(Lifetime other)
```

Comment:
nit: I had a bit of trouble with the concept of "convertibility" and thus the name here. Would `IsWiderThan` work?